### PR TITLE
Unify project forms

### PR DESCRIPTION
**Pull request by jacobwindsor**
dcp-225 and dcp-392

Firstly, apologies about the size of this PR. I couldn't get the install command to work so had to upgrade all of the dependencies and do the relevent angular upgrades. But good news we're now using latest angular :)

A breakdown:

- Upgrades all dependencies to latest major
- Refactors the ProjectForm and ProjectRegistrationForm into a ProjectCreateEdit module
  - Routing for this module is now done within the module (we should do this elsewhere)
  - We only have one form now! The pages for creating and editing just provide wrappers around the forms and add the DOI autofill and caching functionality
  - layout.ts default export has been changed to a function that changes the form layout depending on if wrangler or in create mode
  - **module structure**
    - It's now structured in the same way as the project catalogue (pages and components subdirs). We should start doing this elsewhere
- Fixes some issues that were present with tab changes and enhances the functionality a bit
  - When changing tabs on the project page, the URL updates so that when you refresh, you go back to the same tab
  - When changing tabs on the project sub tab while on the project page, and click "edit", you will be taken to the corresponding tab in the edit page
  - When changing tabs on the project edit page, the URL updates so that when you refresh, you go back to the same tab
- Fixes dcp-392. As far as I could see it was just the data access field that was causing issues
- Changes to use yarn **please use yarn commands now over npm**
- Refactors the ProjectCacheService a bit
  - Justification: I had to refactor to return async values so I could test.
  - I also refactored to return promises instead of observables. This makes more sense in this case as we are dealing with one time, async values (not a stream), and want to execute immediately on execution not on subscription

TODOS:

Listing these as they're a bit out of scope but would be good to do

- Turn the ProjectCreateEdit module into a Project module. Move the project summary page into this module so that we have all project related CRUD ops in one module
- Move several things into the shared module. I put TODOs next to the relevent imports

I know this is a big one so ping me if you want a walkthrough

### CODE REVIEW

File: client/src/app/all-projects/all-projects.component.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export class AllProjectsComponent implements OnInit, OnDestroy {
   @ViewChild(MatPaginator, {static: true}) paginator: MatPaginator;
 
   dataSource: ProjectDataSource;
-  wranglingStates = ingestSchema.properties.wranglingState.enum;
+  wranglingStates = ingestSchema['properties']['wranglingState']['enum'];
```
Comment (amnonkhen) [inline client/src/app/all-projects/all-projects.component.ts:36]: Why keys rather than attributes?

Comment (jacobwindsor) [reply client/src/app/all-projects/all-projects.component.ts:36]: It was causing issues with typescript

File: client/src/app/project-create-edit/components/project-metadata-form/project-metadata-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
+import {Component, EventEmitter, Input, OnDestroy, OnInit, Output, ViewChild} from '@angular/core';
+import * as metadataSchema from '../../schemas/project-metadata-schema.json';
+import * as ingestSchema from '../../schemas/project-ingest-schema.json';
+import {Project} from '../../../shared/models/project';
+import {MetadataFormConfig} from '../../../metadata-schema-form/models/metadata-form-config';
+import {MatTabGroup} from '@angular/material/tabs';
+import {ActivatedRoute, Router} from '@angular/router';
+import {IngestService} from '../../../shared/services/ingest.service';
```
Comment (amnonkhen) [inline client/src/app/project-create-edit/components/project-metadata-form/project-metadata-form.component.ts:8]: relative imports

File: client/src/app/project-create-edit/pages/create-project/create-project.component.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      .subscribe((account) => {
+        this.userIsWrangler = account.isWrangler();
+      });
+  }
+
+  ngOnDestroy() {
+    this.unsubscribe.next();
+  }
+
+  private autofillProjectDetails(id, search): Observable<object> {
```
Comment (amnonkhen) [inline client/src/app/project-create-edit/pages/create-project/create-project.component.ts:56]: Wouldn't this conversion function need a test?

File: client/src/app/project-create-edit/project-create-edit.module.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { ProjectMetadataFormComponent } from './components/project-metadata-form/project-metadata-form.component';
+import { EditProjectComponent } from './pages/edit-project/edit-project.component';
+import { CreateProjectComponent } from './pages/create-project/create-project.component';
```
Comment (amnonkhen) [inline client/src/app/project-create-edit/project-create-edit.module.ts:5]: Please have a consistent spacing in the import statement.

File: client/src/app/project-create-edit/project-create-edit.module.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import { NgModule } from '@angular/core';
```
Comment (amnonkhen) [inline client/src/app/project-create-edit/project-create-edit.module.ts:1]: Please keep import statements sorted.

File: client/src/app/project-create-edit/project-create-edit.module.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import {MatTabsModule} from '@angular/material/tabs';
+import {WranglerListInputComponent} from './components/wrangler-list-input/wrangler-list-input.component';
+import {WranglingPriorityInputComponent} from './components/wrangling-priority-input/wrangling-priority-input.component';
+import {ProjectCacheService} from './services/project-cache.service';
+import {AutofillProjectService} from './services/autofill-project.service';
+import {SharedModule} from '../shared/shared.module';
+
+@NgModule({
+  declarations: [
+    ProjectMetadataFormComponent,
```
Comment (amnonkhen) [inline client/src/app/project-create-edit/project-create-edit.module.ts:27]: Please keep these lists sorted. It helps when merging the file.

File: client/src/app/project/project.component.html [hunk 7]
```diff
@@ -6,12 +6,12 @@
...
   <h1>{{getProjectName()}}</h1>
   <br/>
-  <mat-tab-group [selectedIndex]="lookupTabIndex(selectedProjectTabKey)">
+  <mat-tab-group [selectedIndex]="lookupTabIndex(selectedMainTabKey)" (selectedTabChange)="mainTabChange($event)">
     <mat-tab [label]="tabConfig[0].label">
       <div class="vf-u-margin__top--xxl">
         <app-project-summary (tabChange)="projectTabChange($event)" [project]="project"></app-project-summary>
         <a class="vf-button vf-button--sm {{upload ? '' : 'vf-button--primary'}}" role="button"
-           routerLink="/projects/{{project?.uuid?.uuid}}/{{selectedProjectTabKey}}">Edit Project</a>
+           routerLink="/projects/{{project?.uuid?.uuid}}/edit" [queryParams]="{tab: selectedProjectTabKey}">Edit Project</a>
```
Comment (amnonkhen) [inline client/src/app/project/project.component.html:14]: queryParams attribute in a separate line.

File: client/src/app/shared/services/alert.service.ts [hunk 8]
```diff
@@ -50,6 +50,6 @@ export class AlertService {
 
   clear() {
     // clear alerts
```
Comment (amnonkhen) [inline client/src/app/shared/services/alert.service.ts:52]: Why not rename the function and remove the comment?

**Review by amnonkhen — CHANGES_REQUESTED** [review]

**Comment by amnonkhen** [comment]
Are there any tests that cover this large refactor?

File: client/src/app/metadata-schema-form/custom/select-list/select-list.component.css [hunk 9]
```diff
@@ -12,3 +12,8 @@ div.content, .content.read-only {
   line-height: 1.25em;
   text-align: justify;
 }
+.vf-form__select.force_invalid {
+  border-color:#d32f2f
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply client/src/app/metadata-schema-form/custom/select-list/select-list.component.css:16]: I think this is a good idea but out of scope for this PR

File: client/src/app/shared/services/alert.service.ts (hunk 8 above)
Comment (jacobwindsor) [reply client/src/app/shared/services/alert.service.ts:52]: I think this is out of scope for this PR

**Comment by jacobwindsor** [comment]
@amnonkhen there were no pre-existing tests specifically for the project forms

File: client/src/app/project-create-edit/components/admin-area/admin-area.component.ts [hunk 10]
```diff
@@ -1,7 +1,8 @@
 import {Component, Input, OnInit} from '@angular/core';
 import {FormControl} from '@angular/forms';
-import {Metadata} from '../../metadata-schema-form/models/metadata';
-import {MetadataForm} from '../../metadata-schema-form/models/metadata-form';
+// TODO move these into shared
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply client/src/app/project-create-edit/components/admin-area/admin-area.component.ts:3]: I think this is also out of scope. That's why i left the TODOs

File: client/src/app/project-create-edit/components/accession-field-group/accession-field-group.component.spec.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
 import {async, ComponentFixture, TestBed} from '@angular/core/testing';
 
 import {AccessionFieldGroupComponent} from './accession-field-group.component';
-import {MetadataForm} from '../../metadata-schema-form/models/metadata-form';
+import {MetadataForm} from '../../../metadata-schema-form/models/metadata-form';
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply client/src/app/project-create-edit/components/accession-field-group/accession-field-group.component.spec.ts:4]: I also think this is a good idea but is out of scope

File: client/src/app/project-create-edit/pages/create-project/create-project.component.ts (hunk 3 above)
Comment (jacobwindsor) [reply client/src/app/project-create-edit/pages/create-project/create-project.component.ts:56]: There was no test  when it was implemented previously that I could find. That isn't to say I don't think there should be one just that I wasn't concerned with adding new testing in this PR

File: client/src/app/project-create-edit/project-create-edit.module.ts (hunk 4 above)
Comment (jacobwindsor) [reply client/src/app/project-create-edit/project-create-edit.module.ts:5]: Please see #110. I made a seperte PR as I've optimized imports across all files

File: client/src/app/shared/services/alert.service.ts (hunk 8 above)
Comment (jacobwindsor) [reply client/src/app/shared/services/alert.service.ts:52]: Done but please, let's just use a code formatter. These kind of comments don't need to be part of a PR

File: client/src/app/project/project.component.html (hunk 7 above)
Comment (jacobwindsor) [reply client/src/app/project/project.component.html]: Done but please, let's just use a code formatter. These kind of comments don't need to be part of a PR

File: client/src/app/shared/services/alert.service.ts (hunk 8 above)
Comment (jacobwindsor) [reply client/src/app/shared/services/alert.service.ts]: but done

File: client/src/app/project-create-edit/components/project-metadata-form/project-metadata-form.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    fixture = TestBed.createComponent(ProjectMetadataFormComponent);
+    component = fixture.componentInstance;
+    component.project = {
+      content: {},
+      isInCatalogue: true,
+    };
+    fixture.detectChanges();
+  });
+
+  describe('contributor flow', () => {
```
Comment (amnonkhen) [inline client/src/app/project-create-edit/components/project-metadata-form/project-metadata-form.component.spec.ts:63]: 👍  exactly what I meant.

**Review by amnonkhen — DISMISSED** [review]

File: client/src/app/metadata-schema-form/metadata-schema-form.module.ts [hunk 13]
```diff
@@ -80,7 +80,7 @@ import { VfTabComponent } from './custom/vf-tab/vf-tab.component';
     BaseInputComponent,
     SelectComponent,
     EnumDropDownComponent,
-    SelectListComponent
+    SelectListComponent,
```
Comment (aaclan-ebi) [inline client/src/app/metadata-schema-form/metadata-schema-form.module.ts:83]: clean up dangling comma?

**Review by aaclan-ebi — DISMISSED** [review]
Great work!

File: client/src/app/metadata-schema-form/metadata-schema-form.module.ts (hunk 13 above)
Comment (jacobwindsor) [reply client/src/app/metadata-schema-form/metadata-schema-form.module.ts:83]: I actually like dangling commas. Makes it easier to add something to the list. But we should have a consistent style for this

Comment (aaclan-ebi) [reply client/src/app/metadata-schema-form/metadata-schema-form.module.ts:83]: I see the benefits of dangling commas, yeah. And I read there's no issue with transpilers/bundlers. Ignore my comment.

**Pull request merged**